log-levels.md

```diff
@@ -0,0 +1,203 @@
+log.trace("Finished streaming response", metadata: ["id": "..."]) // good, the same ID is propagated
+```
+
+Thanks to the correlation ID (or a tracing provided ID, in which case we'd log as `context.log.trace("...")` as the ID is propagated automatically), in each following log statement after the initial log statement we're able to correlate all those log statements and know that this `"Finished streaming response"` message was about a response with a `responseCode` that we're able to look up from the `"Received response"` log message.
+
+This pattern is somewhat advanced and may not always be the right approach, but consider it in high performance code where logging the same information repeatedly can be too costly.
+
+##### Things to avoid with Correlation ID logging
+
+When logging with correlation contexts make sure to never "drop the ID", it is easiest to get this right when using distributed tracing's LoggingContext since propagating it ensures the carrying of identifiers, however the same applies to any kind of correlation identifier.
+
+Specifically, avoid situations like these:
+
+```swift
+debug: conection established [connection-id: 7]
+debug: connection closed unexpectedly [error: foobar] // BAD, the connection-id was dropped
+```
+
+On the second line, we don't know which connection had the error since the `connection-id` was dropped. Make sure to audit your logging code to ensure all relevant log statements carry neccessary correlation identifiers.
+
+#### Structured Logging (Semantic Logging)
+
+Libraries may want to embrace the structured logging style. 
+
+It is a fantastic pattern which makes consuming logs in automated systems, and through grepping and other means much easier and future proof.
+
+Consider the folowing not structured log statement:
+
+```swift
+// NOT structured logging style
+log.info("Accepted connection \(connection.id) from \(connection.peer), total: \(connections.count)")
+```
+
+It contains 4 peieces of information:
+
+- we accepted a connection,
+- this is it's string representation,
+- it is from this peer,
+- we currently have `connections.count` active connections.
+
+While this log statement contains all useful information that we meant to relay to end users, it is hard to visually and mechanically parse the detailed information it contains. For example, if we know connections start failing around the time when we reach a total of 100 concurrent connections, it is not trivial to find the specific log statement at which we hit this number. We would have to `grep 'total: 100'` for example, however perhaps there are many other `"total: "` strings present in all of our log systems.
+
+Instead, we can express the same information using the structured logging pattern, as follows:
+
+```swift
+log.info("Accepted connection", metadata: [
+  "connection.id": "\(connnection.id)",
+  "connection.peer": "\(connection.peer)",
+  "connections.total": "\(connections.count)"
```

Comment:
Do we want to give any guidance on the formatting of metadata keys?

camelCase vs. snake_case vs. snake_case but with dots? prefixes for libraries?

Reply:
I currently use `ahc:request-url` (although AHC itself still uses `ahc-request-id` which was my previous preferred style :P), ie. short library ID then colon, then whatever-the-name-of-this-case-is

Reply:
well i could see problems with the colon in some query interfaces. anyway. whatever we decide on, i think we should write this down somewhere, to get some consistency.

Reply:
I'm not sure what various collector systems survive but a typical "key separator" is `.` because those can treat things as "well, basically it's json keys". I'll add it as n idea but i don't think we have a strong recommendation here.